### ISSUE TITLE
ChIPexoQualExample

### ISSUE THREAD

**Issue by welch16** [Bioconductor/Contributions#79]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_
- Repository: https://github.com/welch16/ChIPexoQualExample

Confirm the following by editing each check box to '[x]'
- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.
- [x] I have read the _Bioconductor_ [Package Submission](https://bioconductor.org/developers/package-submission/)
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines](https://bioconductor.org/developers/package-guidelines/).
- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.
- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site](https://support.bioconductor.org) for issues that users may
  have, subscribing to the [bioc-devel](https://stat.ethz.ch/mailman/listinfo/bioc-devel) mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:
- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
    months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion](http://bioconductor.org/developers/how-to/source-control/)
  (optionally [via GitHub](http://bioconductor.org/developers/how-to/git-mirrors/)).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel](https://stat.ethz.ch/mailman/listinfo/bioc-devel) mailing list.

**Comment by welch16** [comment]
Hi @LiNk-NY ,
I hope everything is going great. I only submitted an Experiment package. In the submission guidelines it said that I need to pass the 'moderator' step, is the moderator step you being assigned to review the package or should I need to wait for and fixed your comments to submit the actual package ?
Thanks in advance, Rene

**Comment by welch16** [comment]
AdditionalPackage: https://github.com/welch16/ChIPexoQual

**Comment by LiNk-NY** [comment]
Hi @welch16, 
The moderator step is the package review process. 
It seems like you should also submit the software package along with the data package so that they are both reviewed together.

**Comment by welch16** [comment]
Hi @LiNk-NY ,

Thanks for the explanation and the review. Then, should I trigger another build on **ChIPexoQual**, since the report showed an error since **ChIPexoQualExample** wasn't available?

**Comment by LiNk-NY** [comment]
Hi @welch16, 
You should create another issue and submit `ChIPexoQualExample` software package.

**Comment by LiNk-NY** [comment]
Once the first package builds and passes the build process it should become available for your software package `ChIPexoQual`. I'm not sure why you're getting this error. I will check.

**Comment by lshep** [comment]
Please note there will be an ERROR concerning package dependencies.   

In ChIPexoQual you have specified ChIPexoQualExample (>= 1.00)  in Suggests of the DESCRIPTION file.   

ChIPexoQualExample the current version is Version: 0.99.2

**Comment by welch16** [comment]
Hi @LiNk-NY ,

I saw the review, how can I see the report of the MAC build ? Also, the warning in the linux build occurs due to the size of an example run that I used in the vignette. Should I I generate that run in the vignette?

**Comment by LiNk-NY** [comment]
Hi @welch16, 
It seems like the package is failing on Windows and Mac OS X. 
Were you able to see the report for the Mac build? 

If you're going to include data in `ChIPexoQual`, please be aware that your package cannot be greater than 4 Mb in size after R CMD build. You would have to reduce the size of the example data in `ChIPexoQual`.

**Comment by welch16** [comment]
Hi @LiNk-NY ,

Thanks for all your help. I saw that there was an error and it must be related to a test, but wasn't able to see the specific report in the file :
  /Users/pkgbuild/packagebuilder/workers/jobs/ChIPexoQual/c7fac6a/ChIPexoQual.Rcheck/00check.log

How can I see that file ?

**Comment by LiNk-NY** [comment]
Hi @welch16, 
The `00check.log` contains largely the same information posted in the check output. I will see if this is easily accessible. 
By the way, you can see the line numbers in the files for the tests that are failing.

**Comment by welch16** [comment]
Hi @LiNk-NY ,

I reviewed the report above for version 0.99.4 , and it seems I don't have any warnings or notes from running `R CMD check`. However, in the summary table at the top I still have a warning label. Is that alright?

Also, I reviewed the test in MAC build and find very weird that I have an error in that test but not in others that are similar. Is there a way to have more details in that run ? 

Thanks for the patience

**Comment by LiNk-NY** [comment]
Hi @welch16, 
You have a couple of errors with your build. 
First, you're not using appropriate biocViews for the package. 
See: https://www.bioconductor.org/developers/how-to/biocViews/

Second, there is an error in the vignette. 
```
 ERROR
Quitting from lines 77-88 (vignette.Rmd) 
Error: processing vignette 'vignette.Rmd' failed with diagnostics:
'mc.cores' > 1 is not supported on Windows
Execution halted
```
Third, you are getting an error from the unit tests for `morelia`. 

Regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi @welch16, 
I am briefly looking at your package. 

### ChIPexoQual

#### Grammar 
* The use of underscores and periods are discouraged for function names (including generics and methods). 
* Avoid evaluating multiple expressions in one line with `;` (instead use d <- u <- w <- NULL)
* Try to avoid using obscure column/object names (eg., f, d, r)

**Comment by welch16** [comment]
Hi @LiNk-NY ,

Thanks for your response, 

I don't follow the first point completely. For example, I have this function named `.MA_DF`, I started with a period as this function is internal and that way `roxygen2` will ignore the documentation comments that I added. Should I have added another name as `.MADF` or `.MA_DataFrame` ?

**Comment by LiNk-NY** [comment]
Hi @welch16, 

Our convention is to use **camelCase** on exported functions.  

But why do these functions need to be internal methods or generics in the first place? 

Aren't they just helper functions? You should be able to apply them in the appropriate context within a function. 
It doesn't seem like the way to go to me..
Regards, 
Marcel

**Comment by LiNk-NY** [comment]
It doesn't seem to me like there are many classes that you're dispatching these internal methods on. There would be some value if you had a specific method per signature but it doesn't seem the case.

**Comment by welch16** [comment]
Hi @LiNk-NY ,

I am really sorry if I made your work harder, I updated the package with the fixes you mentioned above. I am not sure why the build hasn't proceed into the bioc system. For the next time, should I wait a bit before making a new build or should I work this changes as you mention them ?

Thanks for your patience

**Comment by lshep** [comment]
Sorry. We were making updates to the builder around the same time as your commit. I will get a new build on for you.

**Comment by LiNk-NY** [comment]
Thanks @lshep! 

@welch16, thank you for making the changes. 
For the functions such as `MADataFrame`, do they need to be exported? Do you think an end-user of your package would need it? Perhaps, this is an intermediate step and could be best written if you use a naming convention such as `.MADataFrame` to avoid exporting the function and to use it as a helper. 

Thank you for your attention. 

Regards,
Marcel

**Comment by welch16** [comment]
@LiNk-NY,
Originally I thought they wont be necessary. But as these function generate the data used for the qc plots that the package generates, I thought they may be used to perform more complex analysis.

Please let me know your thoughts,

**Comment by welch16** [comment]
Hi @LiNk-NY and @lshep,

I pushed a build today and it raised a warning due to the release of the new bioconductor version. To fix it, I installed the future R 3.4, and  increased the required version in the package's DESCRIPTION, it seems that this is raising an error. Could you point me toward to the right way to proceed ?

Thanks in advance,

**Comment by LiNk-NY** [comment]
Hi @lshep, 
Is the SPB running BioC 3.4 or 3.5? 

Thanks.

**Comment by lshep** [comment]
3.4. We are still in the process of moving over the SPB to 3.5. We hope to have it resolved soon.  For now please ignore the warning with the version numbering.

**Comment by LiNk-NY** [comment]
Thanks @lshep! 

@welch16, please lower the R version (to >= 3.3.1) so that it works with BioC 3.4 on the SPB. Also, bump the version for a new build. 

Regards,
Marcel

**Comment by welch16** [comment]
Hi @LiNk-NY ,

Before this issue with the release of Bioconductor 3.4 and the subsequent warnings and errors, we were discussing about exporting or not the functions that generate DataFrames for plots and such. Should I export those functions or not ?

Thanks for your help, Rene

**Comment by LiNk-NY** [comment]
If you think this process is valuable to the user (i.e., the user takes data from a standard file format and usually converts it to the DataFrame shape), then you should export these functions. 
This looks like an intermediate step. Couldn't it be included in the function that creates the ExoData object and specified via arguments? 

--- 

### Considerations for `ChIPexoQual`

**Output from `MangoTheCat/goodpractice::gp`**

* Please use the assignment operator (`<-`) for all object assignments. 
* Remove the date field from the `DESCRIPTION`
* Add a BugReports field in the `DESCRIPTION` (link to github repository)
* Do not import entire packages, try to selectively import the functions that you need in the `NAMESPACE`. You have done some of that already. 

### Considerations for `ChIPexoQualExample`

* Do you have to use exportPattern in the `NAMESPACE`?

**Comment by welch16** [comment]
Hi @LiNk-NY , @lshep 

The label above is a new one. Are you updating your system by any chance ? 

Thanks for the help,
Best, Rene

**Comment by LiNk-NY** [comment]
Hi Rene,  @welch16 
@lshep 
Lori is working on the build system and your package will likely require a manual rebuild. More updates from Lori soon.  

Thanks, 
Marcel

**Comment by lshep** [comment]
The "abnormal" is an error on our end from updating the single package builder. Once the update is complete I will manually kick off a build of your package
I will do a manual rebuild of both your data package and main package to ensure both get on the new build nodes so you may get a few build notices.

**Comment by LiNk-NY** [comment]
Thanks Lori!

**Comment by welch16** [comment]
Hi @lshep , 

I hope you had a great weekend. When I checked the build report of the package I noticed the following error on the windows build.

```
Warning in dir.create(exdir, showWarnings = TRUE, recursive = TRUE) :
  'C:\Windows\TEMP\RtmpmA4sF0' already exists
Error in untar(pkgname, exdir = t) : 
  failed to create directory 'C:\Windows\TEMP\RtmpmA4sF0'
Calls: <Anonymous> -> BiocCheck -> .get_package_dir -> untar
Execution halted
```
It seems that this error is due to permissions or something. Is there anything I could do to fix it ?

Thanks for the help

**Comment by lshep** [comment]
I'm going to try to rerun the build manually to see if it might have just been a problem when we were updating the system.  If it persists I'll investigate further.

**Comment by LiNk-NY** [comment]
Hi René, @welch16
Are you missing an import `RSQLite` directive in the `DESCRIPTION`? 

Thanks,
Marcel 

```
 Error in library.dynam(lib, package, package.lib) : 
    shared object 'RSQLite.so' not found
```

**Comment by welch16** [comment]
Hi Marcel, @LiNk-NY 

I am not sure about it, as `BiocCheck` is working correctly in the systems that I am using. Also, this commit was only to downgrade the versions of `S4Vectors` and `BiovizBase` in the `DESCRIPTION` file. Could it be an issue with `BiocCheck` ?

Thanks, Rene

**Comment by LiNk-NY** [comment]
Hi René, @welch16
I don't get an error when building locally or when checking using `BiocCheck`. Did you bump the version to trigger another build with your latest updates? 
Thanks,
Marcel

**Comment by welch16** [comment]
Hi Marcel, @LiNk-NY ,

I didn't got an error either, that is way I thought it has something to do with the building process. I haven't triggered another build, as I was just creating a static documentation for the package with `pkgdown` and I don't think that is supposed to raise any issues. Should I trigger a new build here ?

Best, Rene

**Comment by LiNk-NY** [comment]
Hi René, @welch16 

Yes, you should trigger a new build. Any changes you make should warrant an update of the version number. 

Regards,
Marcel

**Comment by welch16** [comment]
Hi Marcel @LiNk-NY ,

I triggered a new build and it seems there is an error in the MAC build check, it seems pdflatex is not available in that machine, as the following message appears:

```
LaTeX errors when creating PDF version.
This typically indicates Rd problems.
* checking PDF version of manual without hyperrefs or index ... ERROR
Re-running with no redirection of stdout/stderr.
Hmm ... looks like a package
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in running tools::texi2pdf()
```
Thanks in advance, Rene

**Comment by LiNk-NY** [comment]
Hi René, 
We are looking into this `texi2dvi` error. 
Unfortunately, we need to have a clean build report for the software package to build. 

I will look at it manually. 

Regards,
Marcel

**Comment by LiNk-NY** [comment]
Hi René @welch16, 
Thanks for submitting to Bioconductor. Your package has been approved. 
Regards,
Marcel

**Comment by LiNk-NY** [comment]
Additional comments to keep in mind: 

From `goodpractice::gp("ChIPexoQual")`: 
```
use '<-' for assignment instead of '='. '<-' is the
    standard, and R users and developers are used it and it is easier
    to read your code for them if you use '<-'.

    R/base_summaryPlots.R:173:19
    R/base_summaryPlots.R:404:14
    R/base_summaryPlots.R:471:31
```

Also, please consider adding **unit tests**.
